File: 3534.path-existence-queries-in-a-graph-ii.py
#
# @lc app=leetcode id=3534 lang=python3
#
# [3534] Path Existence Queries in a Graph II
#

# @lc code=start
from typing import List
class Solution:
    def pathExistenceQueries(self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]) -> List[int]:
        # Step 1: Initialize Union-Find structure
        parent = list(range(n))
        rank = [0] * n

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            rootX = find(x)
            rootY = find(y)
            if rootX != rootY:
                if rank[rootX] > rank[rootY]:
                    parent[rootY] = rootX
                elif rank[rootX] < rank[rootY]:
                    parent[rootX] = rootY
                else:
                    parent[rootY] = rootX
                    rank[rootX] += 1
        
        # Step 2 and 3: Sort and construct edges using sliding window technique
        indexed_nums = sorted((num, index) for index, num in enumerate(nums))
        i = 0
        for j in range(n):
            while indexed_nums[j][0] - indexed_nums[i][0] > maxDiff:
                i += 1
            # Union all nodes within current window satisfying condition
            for k in range(i, j):
                union(indexed_nums[k][1], indexed_nums[j][1])
        
        # Step 5: Process each query using Union-Find results
        result = []
        for ui, vi in queries:
            if find(ui) == find(vi):
                result.append(1)
            else:
                result.append(-1)
        return result
# @lc code=end

File: 3534.path-existence-queries-in-a-graph-ii.cpp
# @lc code=start
class Solution {
public:
    vector<int> pathExistenceQueries(int n, vector<int>& nums, int maxDiff, vector<vector<int>>& queries) {
        vector<vector<int>> adj(n);
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                if (abs(nums[i] - nums[j]) <= maxDiff) {
                    adj[i].push_back(j);
                    adj[j].push_back(i);
                }
            }
        }
        vector<int> answer;
        for (const auto& query : queries) {
            int ui = query[0], vi = query[1];
            if (ui == vi) {
                answer.push_back(0);
                continue;
            }
            vector<int> dist(n, -1);
            queue<int> q;
            q.push(ui);
            dist[ui] = 0;
            bool found = false;
            while (!q.empty() && !found) {
                int node = q.front();
                q.pop();
                for (int neighbor : adj[node]) {
                    if (dist[neighbor] == -1) { // not visited
                        dist[neighbor] = dist[node] + 1;
                        if (neighbor == vi) {
                            answer.push_back(dist[neighbor]);
                            found = true;
                            break;
                        }
                        q.push(neighbor);
                    }
                }
            }
            if (!found) answer.push_back(-1);
        }
        return answer;
    }
};
# @lc code=end